File: src/Billing/AutoPurchaseService.php
<?php
declare(strict_types=1);
namespace App\Billing;
use App\Infrastructure\{Database,Outbox};
final class AutoPurchaseService
{
    public function __construct(private Database $db, private Outbox $outbox, private Wallet $wallet, private CartService $carts, private BillingService $billing) {}
    /**
     * After a successful topup, if the user saved a cart with topup intent,
     * charge the balance and complete the purchase. Idempotent.
     */
    public function afterTopup(string $userId): void
    {
        $cart = $this->carts->get($userId);
        if (!$cart || empty($cart['_intent'])) return;
        $this->carts->clearIntent($userId);
        $kind = $cart['kind'] ?? 'subscription';
        try {
            if ($kind === 'subscription') {
                $this->purchaseSubscription($userId, $cart);
            } elseif ($kind === 'gift') {
                $this->purchaseGift($userId, $cart);
            } elseif ($kind === 'traffic') {
                $this->purchaseTraffic($userId, $cart);
            } elseif ($kind === 'devices') {
                $this->purchaseDevices($userId, $cart);
            }
        } catch (BillingError $e) {
            // Keep the cart so the user can retry; notify via outbox.
            $this->carts->save($userId, $cart, true);
            $this->outbox->enqueue('telegram.send', 'cart-failed:' . $userId . ':' . time(), [
                'chat_id' => $this->telegramId($userId),
                'text' => 'Не удалось завершить покупку после пополнения: ' . $e->getMessage() . ' Пополните баланс или повторите покупку в кабинете.',
            ]);
        }
    }
    private function purchaseSubscription(string $userId, array $cart): void
    {
        $planId = (string)($cart['plan_id'] ?? '');
        $key = (string)($cart['idempotency_key'] ?? 'cart:' . $userId . ':' . $planId);
        $plan = $this->db->one('SELECT * FROM plans WHERE id=? AND active=1', [$planId]);
        if (!$plan) throw new BillingError('Тариф больше недоступен.');
        $this->db->transaction(function () use ($userId, $plan, $key) {
            $this->wallet->debit($userId, (int)$plan['price_minor'], 'subscription_purchase', 'Покупка подписки: ' . $plan['name']);
            $order = $this->billing->order($userId, $plan['id'], $key, null, null);
            $this->billing->settleFromBalance($order['id'], (int)$plan['price_minor'], 'RUB');
        });
        $this->carts->delete($userId);
    }
    private function purchaseGift(string $userId, array $cart): void
    {
        $planId = (string)($cart['plan_id'] ?? '');
        $plan = $this->db->one('SELECT * FROM plans WHERE id=? AND active=1', [$planId]);
        if (!$plan) throw new BillingError('Тариф для подарка больше недоступен.');
        $this->db->transaction(function () use ($userId, $plan, $cart) {
            $this->wallet->debit($userId, (int)$plan['price_minor'], 'gift_purchase', 'Покупка подарочной подписки: ' . $plan['name']);
            $this->outbox->enqueue('gift.create', 'gift:' . $userId . ':' . $plan['id'] . ':' . time(), [
                'user_id' => $userId,
                'plan_id' => $plan['id'],
                'price_minor' => (int)$plan['price_minor'],
                'currency' => 'RUB',
            ]);
        });
        $this->carts->delete($userId);
    }
    private function purchaseTraffic(string $userId, array $cart): void
    {
        $subscriptionId = (string)($cart['subscription_id'] ?? '');
        $gb = (int)($cart['traffic_gb'] ?? 0);
        $price = (int)($cart['price_kopeks'] ?? 0);
        if ($gb < 1 || $price < 1) throw new BillingError('Некорректный пакет трафика.');
        $this->db->transaction(function () use ($userId, $subscriptionId, $gb, $price) {
            $sub = $this->db->one('SELECT * FROM subscriptions WHERE id=? AND user_id=? AND status=\'active\'' . $this->db->lock(), [$subscriptionId, $userId]);
            if (!$sub) throw new BillingError('Подписка не найдена.');
            $this->wallet->debit($userId, $price, 'traffic_topup', 'Докупка трафика: ' . $gb . ' ГБ');
            $this->db->execute('UPDATE subscriptions SET purchased_traffic_gb = purchased_traffic_gb + ? WHERE id=?', [$gb, $subscriptionId]);
            $this->outbox->enqueue('subscription.traffic', 'traffic:' . $subscriptionId . ':' . $gb, ['subscription_id' => $subscriptionId, 'traffic_gb' => $gb]);
        });
        $this->carts->delete($userId);
    }
    private function purchaseDevices(string $userId, array $cart): void
    {
        $subscriptionId = (string)($cart['subscription_id'] ?? '');
        $count = (int)($cart['devices'] ?? 0);
        $price = (int)($cart['price_kopeks'] ?? 0);
        if ($count < 1 || $price < 1) throw new BillingError('Некорректное количество устройств.');
        $this->db->transaction(function () use ($userId, $subscriptionId, $count, $price) {
            $sub = $this->db->one('SELECT * FROM subscriptions WHERE id=? AND user_id=? AND status=\'active\'' . $this->db->lock(), [$subscriptionId, $userId]);
            if (!$sub) throw new BillingError('Подписка не найдена.');
            $this->wallet->debit($userId, $price, 'device_addon', 'Докупка устройств: +' . $count);
            $this->db->execute('UPDATE subscriptions SET device_limit = device_limit + ? WHERE id=?', [$count, $subscriptionId]);
            $this->outbox->enqueue('subscription.devices', 'devices:' . $subscriptionId . ':' . $count, ['subscription_id' => $subscriptionId, 'devices' => $count]);
        });
        $this->carts->delete($userId);
    }
    private function telegramId(string $userId): string
    {
        return (string)($this->db->one('SELECT telegram_id FROM users WHERE id=?', [$userId])['telegram_id'] ?? '');
    }
}

File: ruby_app/zeleboba/billing/auto_purchase_service.rb
# frozen_string_literal: true

require_relative "error"

module Zeleboba
  module Billing
    class AutoPurchaseService
      def initialize(db, carts, billing, gifts, wallet, outbox)
        @db = db
        @carts = carts
        @billing = billing
        @gifts = gifts
        @wallet = wallet
        @outbox = outbox
      end

      def after_topup(user_id)
        cart = nil
        @db.transaction do
          cart = @carts.fetch(user_id)
          return false unless cart && cart["_intent"]

          # Clear first. On failure the exact cart is restored, which prevents
          # duplicate completions when a worker is retried after a crash.
          @carts.clear_intent(user_id)
          process(user_id, cart)
          @carts.delete(user_id)
          true
        rescue Error => e
          @carts.save(user_id, cart.reject { |key, _| key == "_intent" }, intent: true) if cart
          notify_failure(user_id, e.message)
          false
        end
      end

      private

      def process(user_id, cart)
        case cart.fetch("kind", "subscription")
        when "subscription"
          @billing.purchase_from_balance(user_id, cart.fetch("plan_id"), key(cart, user_id))
        when "gift"
          @gifts.purchase_from_balance(user_id, cart.fetch("plan_id"), key(cart, user_id), recipient_type: cart["recipient_type"], recipient_value: cart["recipient_value"], message: cart["message"])
        when "traffic"
          purchase_traffic(user_id, cart)
        when "devices"
          purchase_devices(user_id, cart)
        else
          raise Error, "Неизвестный тип отложенной покупки."
        end
      end

      def purchase_traffic(user_id, cart)
        gb = Integer(cart.fetch("traffic_gb"))
        price = Integer(cart.fetch("price_kopeks"))
        raise Error, "Некорректный пакет трафика." unless gb.positive? && price.positive?
        subscription = active_subscription(user_id, cart.fetch("subscription_id"))
        @wallet.debit(user_id, price, "traffic_topup", "Докупка трафика: #{gb} ГБ", "balance", key(cart, user_id))
        @db.execute("UPDATE subscriptions SET purchased_traffic_gb=purchased_traffic_gb+?,updated_at=? WHERE id=?", [gb, Time.now.to_i, subscription["id"]])
        @outbox.enqueue("subscription.traffic", "traffic:#{subscription["id"]}:#{key(cart, user_id)}", { "subscription_id" => subscription["id"], "traffic_gb" => gb })
      end

      def purchase_devices(user_id, cart)
        count = Integer(cart.fetch("devices"))
        price = Integer(cart.fetch("price_kopeks"))
        raise Error, "Некорректное количество устройств." unless count.positive? && price.positive?
        subscription = active_subscription(user_id, cart.fetch("subscription_id"))
        @wallet.debit(user_id, price, "device_addon", "Докупка устройств: +#{count}", "balance", key(cart, user_id))
        @db.execute("UPDATE subscriptions SET device_limit=device_limit+?,updated_at=? WHERE id=?", [count, Time.now.to_i, subscription["id"]])
        @outbox.enqueue("subscription.devices", "devices:#{subscription["id"]}:#{key(cart, user_id)}", { "subscription_id" => subscription["id"], "devices" => count })
      end

      def active_subscription(user_id, subscription_id)
        @db.one("SELECT * FROM subscriptions WHERE id=? AND user_id=? AND status='active'#{@db.lock}", [subscription_id, user_id]) || raise(Error, "Подписка не найдена.")
      end

      def key(cart, user_id)
        value = cart["idempotency_key"].to_s
        raise Error, "Некорректный ключ операции." unless value.match?(/\A[a-zA-Z0-9:_-]{8,128}\z/)

        value
      end

      def notify_failure(user_id, message)
        telegram_id = @db.one("SELECT telegram_id FROM users WHERE id=?", [user_id])&.fetch("telegram_id", nil)
        return if telegram_id.to_s.empty?

        @outbox.enqueue("telegram.send", "cart-failed:#{user_id}:#{Time.now.to_i}", { "chat_id" => telegram_id, "text" => "Не удалось завершить покупку после пополнения: #{message}" })
      end
    end
  end
end
